src/firestore-adapter.ts

```diff
@@ -0,0 +1,859 @@
+      snapshot.docChanges().forEach((change) => {
+        if (change.type === "added") {
+          this._historyChildAdded(change.doc);
+        }
+      });
+    });
+
+    historyRef.get().then(() => {
+      this._handleInitialRevisions();
+    });
+  }
+
+  /**
+   * Apply all pending changes in `history` node that aren't yet checked in into `checkpoint`, and then mark connection to be Ready.
+   */
+  protected _handleInitialRevisions(): void {
+    if (this._zombie) {
+      // just in case we were cleaned up before we got the data.
+      return;
+    }
+
+    Utils.validateFalse(this._ready, "Should not be called multiple times.");
+
+    if (!this._initialRevisions) {
+      this._initialRevisions = true;
+      this._trigger(FirebaseAdapterEvent.InitialRevision);
+    }
+
+    // Compose the checkpoint and all subsequent revisions into a single operation to apply at once.
+    this._revision = this._checkpointRevision;
+
+    let revisionId = this._revisionToId(this._revision);
+    const pending = this._pendingReceivedRevisions;
+
+    while (pending[revisionId] != null) {
+      const revision: IRevision | null = this._parseRevision(
+        pending[revisionId] as any
+      );
+
+      if (!revision) {
+        // If a misbehaved client adds a bad operation, just ignore it.
+        console.log(
+          "Invalid operation.",
+          this._firestoreUserRef!.toString(),
+          revisionId,
+          pending[revisionId]
+        );
+      } else {
+        this._document = this._document!.compose(revision.operation);
+      }
+
+      delete pending[revisionId];
+      this._revision++;
+
+      revisionId = this._revisionToId(this._revision);
+    }
+
+    this._trigger(FirebaseAdapterEvent.Operation, this._document!);
+    this._ready = true;
+
+    setTimeout(() => {
+      this._trigger(FirebaseAdapterEvent.Ready, true);
+    });
+  }
+
+  /**
+   * Apply incoming changes from newly added child in `history` node of Firebase ref.
+   */
+  protected _handlePendingReceivedRevisions(): void {
+    const pending = this._pendingReceivedRevisions;
+
+    let revisionId = this._revisionToId(this._revision);
+    let triggerRetry = false;
+
+    while (pending[revisionId] != null) {
+      this._revision++;
+
+      const revision: IRevision | null = this._parseRevision(
+        pending[revisionId]
+      );
+      if (!revision) {
+        // If a misbehaved client adds a bad operation, just ignore it.
+        console.log(
+          "Invalid operation.",
+          this._firestoreRef!.toString(),
+          revisionId,
+          pending[revisionId]
+        );
+      } else {
+        this._document = this._document!.compose(revision.operation);
+
+        if (this._sent && revisionId === this._sent.id) {
+          // We have an outstanding change at this revision id.
+          if (
+            this._sent.op.equals(revision.operation) &&
+            revision.author == this._userId
```

Comment:
Can we add the check from this PR (https://github.com/interviewstreet/firepad-x/pull/69) here?

Reply:
Lets first go ahead with firestore-only changes, and have a separate release for this change.